content/article/my-perl-wishlist-invariant-sigils-part-1.md

```diff
@@ -0,0 +1,181 @@
+
+Make the sigils part of the name
+--------------------------------
+
+To save myself from repeating my errors, I'd like the sigil to be part of a
+variable's name.  Scalars work this way in Perl 5, bash, and
+[Raku](https://docs.perl6.org/language/101-basics#sigil_and_identifier)
+([formerly Perl 6](https://github.com/perl6/problem-solving/blob/master/solutions/language/Path-to-Raku.md)).
+That would make the above table look like:
+
+| Sigil | No brackets                                     | `[ ]` (array access)                | `{ }` (hash access)                        |
+|-------|-------------------------------------------------|-------------------------------------|--------------------------------------------|
+| `$`   | `$z`: a scalar, i.e., a single value            | `$z[0]`: N/A                        | `$z{0}`: N/A                               |
+| `@`   | `@z`: An array, i.e., a list of value(s)        | `@z[0]`: the first element of `@z`  | `@z{0}`: N/A                               |
+| `%`   | `%z`: A hash, i.e., a list of key/value pair(s) | `%z[0]`: N/A                        | `%z{0}`: the value in hash `%z` at key `0` |
+
+Simpler!
+
+But what about slices?
+----------------------
+
+Slices such as `@z[0,1]` and `%z{qw(hello there)}` return multiple
+values from an array or hash.  If sigils `@` and `%` are no longer
+available for slicing, we need an alternative.
+The Perl community currently provides two models: postderef syntax and
+postfix adverbs.
+
+Perl 5.20+ support
+[postfix dereferencing](https://www.effectiveperlprogramming.com/2014/09/use-postfix-dereferencing/)
+("postderef"), which gives us one option.  Postderef separates the name
+from the slice:
+
+```perl
+# Valid Perl 5.20+
+$hashref->{a};      # Scalar, element at index "a" of the hash pointed to by $hashref
+$hashref->@{a};     # List including the "a" element of the hash pointed to by $hashref
+$hashref->%{a};     # List including the key "a" and the "a" element of the hash pointed to by $hashref
+```
+
+The type of slice comes after the reference, instead of as a sigil
+before the reference.  With non-references, that would give us slices
+such as `@z@[1,2,3]` or `%hash%{a}`.
+
+Raku gives us another option: "adverbs" such as
+[`:kv`](https://docs.perl6.org/language/subscripts#:kv).  For example:
+
+```perl6
+# Valid Raku
+%hash{"a"}          # Single value, element at index "a" of %hash
+%hash{"a"}:v;       # The same --- just the value
+%hash{"a"}:kv;      # The list including key "a" and the value of the "a" element of %hash
+```
+
+The adverb (e.g., `:kv`) goes in postfix position, immediately
+after the brackets or braces.  Following this model,
+slices would look like `@z[1,2,3]:l` or `%hash{a}:kv`.  (I propose `:l`
```

Comment:
It *might* help to reword this slightly to be:

> I propose `:l` as in list instead of `:v` which can return a scalar or a list

Reply:
Revised